### PR TITLE
Printing important decomposition information

### PR DESCRIPTION
**Pull request by tjof2**
### Description of the change
@thomasaarholt and I were discussing the idea of printing out information about `s.decomposition()`, after I discovered that the issue described in #1361 was likely caused by a hidden default argument passed to sklearn.

This PR introduces a `print_info=True/False` argument to `s.decomposition`, which prints out the values of the main arguments of the function such as `normalize_poissonian_noise`, and, if the chosen algorithm is an sklearn estimator, such as NMF, prints out `repr(estimator)` as well to highlight the (default) arguments sklearn is be using.

There are couple of bits open for review here

1. `polyfit=` is not documented in the docstring (see #1159)
2. Should the default value of `print_info` be True or False?
3. Should we log **ALL** the arguments? I've ignored the mask arguments, which can be ndarrays, for example.

### Progress of the PR
- [x] Change implemented
- [x] update docstring (if appropriate)
- [x] update user guide (if appropriate)
- [x] add tests
- [x] ready for review.

### Example
Default arguments/algorithm (no sklearn)
```python
>>> import numpy as np
>>> import hyperspy.api as hs
>>> s0 = hs.signals.Signal2D(np.abs(np.random.randn(16, 16, 512)))
>>> s0.decomposition(output_dimension=3)

Decomposition info:
  normalize_poissonian_noise=False
  algorithm=svd
  output_dimension=3
  centre=None
```
With an sklearn object
```python
>>> import numpy as np
>>> import hyperspy.api as hs
>>> s0 = hs.signals.Signal2D(np.abs(np.random.randn(16, 16, 512)))
>>> s0.decomposition(output_dimension=3, algorithm='nmf')

Decomposition info:
  normalize_poissonian_noise=False
  algorithm=nmf
  output_dimension=3
  centre=None
scikit-learn estimator:
NMF(alpha=0.0, beta_loss='frobenius', init=None, l1_ratio=0.0, max_iter=200,
    n_components=3, random_state=None, shuffle=False, solver='cd', tol=0.0001,
    verbose=0)
```

### CODE REVIEW

**Comment by thomasaarholt** [comment]
This is great! I'd like someone else to chime in on whether this should be on by default. An argument for is visibility of parameters, arguments against include clutter and possible confusion/"decomposition appears scary".

I think the combination of "our" arguments with the sklearn ones after looks very good.

Some parameters are still not covered, like the ones reported in the warning in this screenshot (method, etc.):
```python
import hyperspy.api as hs
import numpy as np
s = hs.signals.Signal1D(np.random.random((64, 64, 100)))
s.decomposition(algorithm='ORPCA', output_dimension=4)
```
image

**Comment by tjof2** [comment]
@thomasaarholt ORPCA is an inconsistency, well spotted!

I think I'd like #2035 merged first before fixing it though - because it requires a little bit of work to e.g. print out the info in the same way as an sklearn estimators via __repr__.

> and possible confusion/"decomposition appears scary"

I'd be more worried that people were using decomposition without understanding why! But a valid point, clarity and brevity is super important to make this effective.

**Comment by tjof2** [comment]
> and possible confusion/"decomposition appears scary"

I'd be more worried that people were using decomposition without understanding why! But a valid point, clarity and brevity is super important to make this effective.

**Comment by tjof2** [comment]
@francisco-dlp  any view on whether this should be "on" or "off" by default?

**Comment by francisco-dlp** [comment]
This is a really nice feature that will help many users discover the many parameters involved in ``decomposition``. Therefore +1 for having this on by default. However, I wonder if this should go to the ``logger`` instead of ``stdout`` as at present.  What do you think @ericpre?

**Comment by tjof2** [comment]
I ~also think the logger is better, and~ agree with it being on by default as you write. I'll tidy this up tomorrow.

That said, this is really INFO logging level information, so 

1. there's then no need for a `print_info` argument, and 
2. if the default logging level is WARNING, the people who would benefit most from this (end-users) won't see this info.

**Comment by thomasaarholt** [comment]
I agree with @tjof2. Since this is INFO, people won't see it. And if it is WARNING, people will worry and cite "errors", when this is nothing to worry about. My vote goes to `print_info=True` by default, since expert users / users running this in a loop will know to turn it off.

**Comment by ericpre** [comment]
Putting the information to stdout or the logger can be important for downstream library or programme. Since the output to stdout can be switch off using `print_info=False`, this should not be an issue.
In this specific case, sending the information is `stdout` is appropriate to my opinion (since it can be disable) and I agree that the default should be `print_info=True`.

File: hyperspy/learn/mva.py [hunk 1]
```diff
@@ -166,6 +173,17 @@ def decomposition(self,
...
         to_return = None
+        to_print = [
+            "Decomposition info:",
+            "  normalize_poissonian_noise={}".format(normalize_poissonian_noise),
+            "  algorithm={}".format(algorithm),
+            "  output_dimension={}".format(output_dimension),
+            "  centre={}".format(centre),
+            "  auto_transpose={}".format(auto_transpose),
+            "  polyfit={}".format(polyfit),
+            "  reproject={}".format(reproject)
```
Comment (ericpre) [inline hyperspy/learn/mva.py]: I think we could remove these to make it more succinct.

Comment (tjof2) [reply hyperspy/learn/mva.py]: I think auto-transpose is important. Agree on the others.

Comment (tjof2) [reply hyperspy/learn/mva.py]: That said, `auto_transpose` only affects SVD/fast SVD, so I'll remove it. I initially thought it affected the sklearn estimators but it doesn't.

**Comment by tjof2** [comment]
Thanks for the comments @ericpre - I've removed some of the variables we print to focus on the important ones, and also added some tests to boost coverage.

Output now looks like this:
```
>>> import numpy as np
>>> import hyperspy.api as hs
>>> s0 = hs.signals.Signal2D(np.abs(np.random.randn(16, 16, 512)))
>>> s0.decomposition(output_dimension=3, algorithm='nmf')

Decomposition info:
  normalize_poissonian_noise=False
  algorithm=nmf
  output_dimension=3
  centre=None
scikit-learn estimator:
NMF(alpha=0.0, beta_loss='frobenius', init=None, l1_ratio=0.0, max_iter=200,
    n_components=3, random_state=None, shuffle=False, solver='cd', tol=0.0001,
    verbose=0)
```

I think this is complete now pending CI tests.

**Comment by tjof2** [comment]
The failures in CI look to be unrelated - and are in `test_image.py`. Perhaps this is related to #2348?

```
hyperspy/tests/mva/test_decomposition.py ................................ [ 56%]
```

**Comment by ericpre** [comment]
> The failures in CI look to be unrelated - and are in `test_image.py`. Perhaps this is related to #2348?
> 
> ```
> hyperspy/tests/mva/test_decomposition.py ................................ [ 56%]
> ```

Yes, the failures are unrelated to this PR.

**Comment by ericpre** [comment]
Once the tests of you last commit are passed on appveyor, we can merged it. Appveyor is doing one build at a time, so it is a bit slow...

**Comment by tjof2** [comment]
You're right @ericpre  it is quite slow! :D getting there though...

Success! :D
